phoenix-core/src/main/java/org/apache/phoenix/expression/function/CollationKeyFunction.java

```diff
@@ -0,0 +1,230 @@
+		// collator strength
+		@FunctionParseNode.Argument(allowedTypes = { PInteger.class }, defaultValue = "null", isConstant = true),
+		// collator decomposition
+		@FunctionParseNode.Argument(allowedTypes = { PInteger.class }, defaultValue = "null", isConstant = true) })
+public class CollationKeyFunction extends ScalarFunction {
+
+	private static final Log LOG = LogFactory.getLog(CollationKeyFunction.class);
+
+	public static final String NAME = "COLLATION_KEY";
+
+	private Collator collator;
+
+	public CollationKeyFunction() {
+	}
+
+	public CollationKeyFunction(List<Expression> children) throws SQLException {
+		super(children);
+		initialize();
+	}
+
+	@Override
+	public void readFields(DataInput input) throws IOException {
+		super.readFields(input);
+		initialize();
+	}
+
+	@Override
+	public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+		try {
+			String inputValue = getInputString(tuple, ptr);
+			byte[] collationKeyByteArray = collator.getCollationKey(inputValue).toByteArray();
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Collation key bytes: " + VarBinaryFormatter.INSTANCE.format(collationKeyByteArray));
+			}
+
+			ptr.set(collationKeyByteArray);
+			return true;
+		} catch (ExpressionEvaluationException e) {
+			LOG.debug("ExpressionEvaluationException caught: " + e.getMessage());
+			return false;
+		}
+	}
+
+	private void initialize() {
+		String localeISOCode = getLiteralValue(1, String.class);
+		Boolean useSpecialUpperCaseCollator = getLiteralValue(2, Boolean.class);
+		Integer collatorStrength = getLiteralValue(3, Integer.class);
+		Integer collatorDecomposition = getLiteralValue(4, Integer.class);
+
+		if (LOG.isDebugEnabled()) {
+			StringBuilder logInputsMessage = new StringBuilder();
+			logInputsMessage.append("Input (literal) arguments:").append("localeISOCode: " + localeISOCode)
+					.append(", useSpecialUpperCaseCollator: " + useSpecialUpperCaseCollator)
+					.append(", collatorStrength: " + collatorStrength)
+					.append(", collatorDecomposition: " + collatorDecomposition);
+			LOG.debug(logInputsMessage);
+		}
+
+		Locale locale = LocaleUtils.get().getLocaleByIsoCode(localeISOCode);
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(String.format("Locale: " + locale.toLanguageTag()));
+		}
+
+		LinguisticSort linguisticSort = LinguisticSort.get(locale);
+
+		collator = BooleanUtils.isTrue(useSpecialUpperCaseCollator) ? linguisticSort.getUpperCaseCollator(false)
+				: linguisticSort.getCollator();
+
+		if (collatorStrength != null) {
+			collator.setStrength(collatorStrength);
+		}
+
+		if (collatorDecomposition != null) {
+			collator.setDecomposition(collatorDecomposition);
+		}
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(String.format("Collator: [strength: %d, decomposition: %d], Special-Upper-Case: %s",
+					collator.getStrength(), collator.getDecomposition(),
+					BooleanUtils.isTrue(useSpecialUpperCaseCollator)));
+		}
+	}
```

Comment:
As of PHOENIX-4294 (just make sure you've pulled the latest), you can declare your function as not being thread safe by adding the following method override:

    @Override
    public boolean isThreadSafe() {
        return false;
    }